File: 04-Data-Structures/code/typescript/src/bloomFilter.ts
import { DSError } from "./errors.js";
import { bloomIndices } from "./hash.js";

/**
 * Classic Bloom filter: `mBits`-wide bit array, `kHashes` independent probe
 * positions per item via FNV-1a double-hashing (see `hash.ts`). Never false
 * negative; may false positive. Hash scheme is identical to the Python port
 * bit-for-bit, so both languages agree on every membership query.
 */
export class BloomFilter {
  private bits: Uint8Array;
  private m: number;
  private k: number;

  constructor(mBits: number, kHashes: number) {
    if (mBits <= 0 || kHashes <= 0) throw new DSError("invalid", "mBits and kHashes must be > 0");
    this.m = mBits;
    this.k = kHashes;
    this.bits = new Uint8Array(mBits);
  }

  add(item: string): void {
    for (const idx of bloomIndices(item, this.k, this.m)) this.bits[idx] = 1;
  }

  mightContain(item: string): boolean {
    for (const idx of bloomIndices(item, this.k, this.m)) {
      if (this.bits[idx] === 0) return false;
    }
    return true;
  }

  checkInvariants(): void {
    if (this.bits.length !== this.m) throw new DSError("invalid", "bit array length mismatch");
  }
}

File: 04-Data-Structures/code/python/seb_ds/bloom_filter.py
"""Classic Bloom filter. Mirrors `typescript/src/bloomFilter.ts`.

Hash scheme is identical to the TypeScript port bit-for-bit (see `hash.py`),
so both languages agree on every membership query for the same inputs.
"""

from __future__ import annotations

from .errors import DSError
from .hash import bloom_indices


class BloomFilter:
    def __init__(self, m_bits: int, k_hashes: int) -> None:
        if m_bits <= 0 or k_hashes <= 0:
            raise DSError("invalid", "mBits and kHashes must be > 0")
        self._m = m_bits
        self._k = k_hashes
        self._bits = bytearray(m_bits)

    def add(self, item: str) -> None:
        for idx in bloom_indices(item, self._k, self._m):
            self._bits[idx] = 1

    def might_contain(self, item: str) -> bool:
        for idx in bloom_indices(item, self._k, self._m):
            if self._bits[idx] == 0:
                return False
        return True

    def check_invariants(self) -> None:
        if len(self._bits) != self._m:
            raise DSError("invalid", "bit array length mismatch")
